src/power_grid_model_io/converters/pandapower_converter.py

```diff
@@ -0,0 +1,460 @@
+
+        # Select the appropriate switches and columns
+        pp_switches = self.pp_data["switch"]
+        pp_switches = pp_switches[self.pp_data["switch"]["et"] == "l"]
+        pp_switches = pp_switches[["element", "bus", "closed"]]
+
+        # Join the switches with the lines twice, once for the from_bus and once for the to_bus
+        pp_from_switches = (
+            pp_lines[["index", "from_bus"]]
+            .merge(
+                pp_switches,
+                how="left",
+                left_on=["index", "from_bus"],
+                right_on=["element", "bus"],
+            )
+            .fillna(True)
+            .set_index(pp_lines.index)
+        )
+        pp_to_switches = (
+            pp_lines[["index", "to_bus"]]
+            .merge(
+                pp_switches,
+                how="left",
+                left_on=["index", "to_bus"],
+                right_on=["element", "bus"],
+            )
+            .fillna(True)
+            .set_index(pp_lines.index)
+        )
+
+        pgm_lines = initialize_array(data_type="input", component_type="line", shape=len(pp_lines))
+        pgm_lines["id"] = self._generate_ids("line", pp_lines.index)
+        pgm_lines["from_node"] = self._get_ids("bus", pp_lines["from_bus"])
+        pgm_lines["from_status"] = pp_lines["in_service"] & pp_from_switches["closed"]
+        pgm_lines["to_node"] = self._get_ids("bus", pp_lines["to_bus"])
+        pgm_lines["to_status"] = pp_lines["in_service"] & pp_to_switches["closed"]
+        pgm_lines["r1"] = pp_lines["r_ohm_per_km"] * pp_lines["length_km"]
+        pgm_lines["x1"] = pp_lines["x_ohm_per_km"] * pp_lines["length_km"]
+        pgm_lines["c1"] = pp_lines["c_nf_per_km"] * pp_lines["length_km"]
+        pgm_lines["tan1"] = 0.0
+        pgm_lines["i_n"] = pp_lines["max_i_ka"] * 1e3
+
+        self.pgm_data["line"] = pgm_lines
+
+    def _create_pgm_input_sources(self):
+        assert "source" not in self.pgm_data
+
+        pp_ext_grid = self.pp_data["ext_grid"]
+
+        pgm_sources = initialize_array(data_type="input", component_type="source", shape=len(pp_ext_grid))
+        pgm_sources["id"] = self._generate_ids("ext_grid", pp_ext_grid.index)
+        pgm_sources["node"] = self._get_ids("bus", pp_ext_grid["bus"])
+        pgm_sources["status"] = pp_ext_grid["in_service"]
+        pgm_sources["u_ref"] = pp_ext_grid["vm_pu"]
+
+        self.pgm_data["source"] = pgm_sources
+
+    def _create_pgm_input_shunts(self):
+        assert "shunt" not in self.pgm_data
+
+        pp_shunt = self.pp_data["shunt"]
+
+        pgm_shunts = initialize_array(data_type="input", component_type="shunt", shape=len(pp_shunt))
+        pgm_shunts["id"] = self._generate_ids("shunt", pp_shunt.index)
+        pgm_shunts["node"] = self._get_ids("bus", pp_shunt["bus"])
+        pgm_shunts["status"] = pp_shunt["in_service"]
+        pgm_shunts["g1"] = (pp_shunt["p_mw"] * 1e6) / ((pp_shunt["vn_kv"] * 1e3) * (pp_shunt["vn_kv"] * 1e3))
```

Comment:
You could combine the multiplications. Right now you have

```python
g = p * 1e6 / ((v * 1e3) * (v * 1e3))
b = q * 1e6 / ((v * 1e3) * (v * 1e3))
```
Which is 10 x N multiplications, while this is the same as

```python
g = p * 1e6 / (v * v * 1e6)
b = q * 1e6 / (v * v * 1e6)
```
Which is 8 x N multiplications, while this is the same as
```python
g = p * v * v
b = q * v * v
```
Which is 4 x N multiplications, while this is the same as
```python
v2 = v * v
g = p * v2
b = q * v2
```
Which is 3 x N multiplications

Reply:
Thank you for such a clear explanation!
A quick question, shouldn't this be division? As in 
g = p / v2
b = q / v2